sydent/http/servlets/registerservlet.py

```diff
@@ -56,22 +59,34 @@ async def render_POST(self, request: Request) -> JsonDict:
                 "error": "matrix_server_name must be a valid Matrix server name (IP address or hostname)",
             }
 
-        result = await self.client.get_json(
-            "matrix://%s/_matrix/federation/v1/openid/userinfo?access_token=%s"
-            % (
-                matrix_server,
-                urllib.parse.quote(args["access_token"]),
-            ),
-            1024 * 5,
-        )
+        try:
+            result = await self.client.get_json(
+                "matrix://%s/_matrix/federation/v1/openid/userinfo?access_token=%s"
+                % (
+                    matrix_server,
+                    urllib.parse.quote(args["access_token"]),
+                ),
+                1024 * 5,
+            )
+        except (DNSLookupError, ConnectError, ResponseFailed) as e:
+            logger.warning("Unable to contact %s: %s", matrix_server, e)
+            request.setResponseCode(HTTPStatus.INTERNAL_SERVER_ERROR)
+            return {
+                "errcode": "M_UNKNOWN",
+                "error": f"Unable to contact the Matrix homeserver ({type(e).__name__})",
+            }
 
         if "sub" not in result:
-            raise Exception("Invalid response from homeserver")
+            request.setResponseCode(HTTPStatus.INTERNAL_SERVER_ERROR)
+            return {
+                "errcode": "M_UNKNOWN",
+                "error": "The Matrix homeserver did not include 'sub' in its response",
+            }
```

Comment:
I think clients don't usually expect JSON bodies for 500 errors so something feels off to me?

I guess this is similar to what we had before but now we'll not error to sentry?

Do we want to log a warning or anything in this case?

Reply:
> I think clients don't usually expect JSON bodies for 500 errors so something feels off to me?

Very possibly. I'm just trying to replicate what we had before (and also try to be a good citizen and provide error context).

> I guess this is similar to what we had before but now we'll not error to sentry?

That's right. I didn't think it merited a sentry warning since it's not a problem in our application---it's duff data from the far end.

I think a warning would be prudent so that we'll have something in the logs to diagnose this. I added something for this on line 72. I'll add something for this and the cases below.